File: cpp/tests/07_grouped_shorts.cpp
#include "test.hpp"

TEST_CASE("grouped short options", "[grouped]") {
  // long options cannot be grouped, so those dont apply
  vector<string> argv = {
    "/test07",
  };
  {
    vector<string> options = GENERATE(
      (vector<string>{"-abcd"}),
      (vector<string>{"-ab", "-cd"}),
      (vector<string>{"-abc", "-d"}),
      (vector<string>{"-abc", "--delta"}),
      (vector<string>{"-ab", "--gamma", "-d"}),
      (vector<string>{"-ac", "--beta", "--delta"}),
      (vector<string>{"--alpha", "-bcd"})
    );
    argv.insert(argv.end(), std::move_iterator(options.begin()), std::move_iterator(options.end()));
  } // destroy options here, it's elements are "moved from"

  CheckArg ca(argv, "test07");
	ca.add('a', "alpha",               "opt a");
	ca.add('b', "beta",    [](auto, const auto&, const auto&) -> int {return 0;}, "opt b");
	ca.add('c', "gamma",               "opt gamma");
	ca.add('d', "delta",   [](auto, const auto&, const auto&) -> int {return 0;}, "opt delta");
	ca.add('e', "epsy",                "opt e");

  int rc = ca.parse();

  INFO("argv:" << argv);
  CHECK(rc == CA_ALLOK);

  REQUIRE(ca.isset("alpha"));
  REQUIRE(ca.isset("beta"));
  REQUIRE(ca.isset("gamma"));
  REQUIRE(ca.isset("delta"));
  REQUIRE(!ca.isset("epsy"));
}

File: rust/tests/07_grouped_shorts.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2021-2022 brainpower <brainpower at mailbox dot org>

#![feature(assert_matches)]
#![feature(get_mut_unchecked)]

#[cfg(test)]
mod tests {
  use std::assert_matches::assert_matches;
  // use std::rc::Rc;
  // use itertools::Itertools;

  use checkarg::{CheckArg, ValueType, RC};

  #[test]
  fn grouped_short_options() {
    // long options cannot be grouped, so those dont apply
    let argvs = vec![
      vec!["/test07", "-abcd"],
      vec!["/test07", "-ab", "-cd"],
      vec!["/test07", "-abc", "-d"],
      vec!["/test07", "-abc", "--delta"],
      vec!["/test07", "-ab", "--gamma", "-d"],
      vec!["/test07", "-ac", "--beta", "--delta"],
      vec!["/test07", "--alpha", "-bcd"],
    ];

    let mut ca = CheckArg::new("test07");
    ca.add('a', "alpha", "opt a", ValueType::None, None);
    ca.add_cb(
      'b',
      "beta",
      "opt b",
      |_, _, _| Ok(()),
      ValueType::None,
      None,
    );
    ca.add('c', "gamma", "opt gamma", ValueType::None, None);
    ca.add_cb(
      'd',
      "delta",
      "opt delta",
      |_, _, _| Ok(()),
      ValueType::None,
      None,
    );
    ca.add('e', "epsy", "opt e", ValueType::None, None);

    for argv in argvs {
      let ret = ca.parse(&argv);

      assert_matches!(ret, RC::Ok);

      assert!(ca.isset("alpha"));
      assert!(ca.isset("beta"));
      assert!(ca.isset("gamma"));
      assert!(ca.isset("delta"));
      assert!(!ca.isset("epsy"));
    }
  }
}
